pyccel/ast/internals.py

```diff
@@ -262,3 +267,31 @@ def symbols(names):
     symbols = [PyccelSymbol(name.strip()) for name in names]
     return tuple(symbols)
 
+def max_precision(objs : list, dtype = None, allow_native = True):
+    """
+    Returns the largest precision of an object in the list
+
+    Parameters
+    ----------
+    objs : list
+           A list of PyccelAstNodes
+    dtype : Dtype class
+            If this argument is provided then only the
+            precision of objects with this dtype are
+            considered
+    """
+    if allow_native and all(o.precision == -1 for o in objs):
+        return -1
+    elif dtype:
+        def_prec = default_precision[str(dtype)]
+        return max(def_prec if o.precision == -1 \
+                else o.precision for o in objs if o.dtype is dtype)
+    else:
+        return max(default_precision[str(o.dtype)] if o.precision == -1 \
+                else o.precision for o in objs)
+
+def get_final_precision(obj):
+    """
+    Get the usable precision of an object
+    """
+    return default_precision[str(obj.dtype)] if obj.precision == -1 else obj.precision
```

Comment:
Please expand the docstring of this function. What do you mean with "final" and "usable" precision?

Reply:
I mean the precision that you can use to print, ie 8 instead of -1 for a default precision `float`. I'm not sure how to express this. Should I just describe the if statement in the docstring?

Reply:
Thanks for the explanation. I would add it verbatim to the docstring, as well as a description of the if statement as you suggested.

Since this function is used all over the library, it is good to be verbose! 😉

Reply:
Fixed : bbb4598